Client/WebFilters/FilterListResourceDownloader.swift

```diff
@@ -0,0 +1,483 @@
+        // This happens because we care about UI changes but not when our downloads finish
+        return
+      }
+        
+      allFilterListSettings[index].isEnabled = isEnabled
+      allFilterListSettings[index].componentId = componentId
+      FilterListSetting.save(inMemory: inMemory)
+    }
+    
+    /// Create a filter list setting for the given UUID and enabled status
+    private func create(uuid: String, componentId: String?, isEnabled: Bool) {
+      let setting = FilterListSetting.create(uuid: uuid, componentId: componentId, isEnabled: isEnabled, inMemory: inMemory)
+      allFilterListSettings.append(setting)
+    }
+  }
+  /// A shared instance of this class
+  ///
+  /// - Warning: You need to wait for `DataController.shared.initializeOnce()` to be called before using this instance
+  public static let shared = FilterListResourceDownloader()
+  
+  /// Object responsible for getting component updates
+  private var adBlockService: AdblockService?
+  /// Manager that handles updates to filter list settings in core data
+  private let settingsManager: FilterListSettingsManager
+  /// The resource downloader that downloads our resources
+  private let resourceDownloader: ResourceDownloader
+  /// The filter list subscription
+  private var filterListSubscription: AnyCancellable?
+  /// Fetch content blocking tasks per filter list
+  private var fetchTasks: [ResourceDownloader.Resource: Task<Void, Error>]
+  /// Ad block service tasks per filter list UUID
+  private var adBlockServiceTasks: [String: Task<Void, Error>]
+  /// A marker that says if fetching has started
+  private var startedFetching = false
+  /// The filter lists wrapped up so we can contain
+  @Published var filterLists: [FilterList]
+  
+  /// A formatter that is used to format a version number
+  private lazy var fileVersionDateFormatter: DateFormatter = {
+    let dateFormatter = DateFormatter()
+    dateFormatter.locale = Locale(identifier: "en_US_POSIX")
+    dateFormatter.dateFormat = "yyyy.MM.dd.HH.mm.ss"
+    dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+    return dateFormatter
+  }()
+  
+  init(networkManager: NetworkManager = NetworkManager(), persistChanges: Bool = true) {
+    self.resourceDownloader = ResourceDownloader(networkManager: networkManager)
+    self.settingsManager = FilterListSettingsManager(inMemory: !persistChanges)
+    self.filterLists = []
+    self.fetchTasks = [:]
+    self.adBlockServiceTasks = [:]
+    self.adBlockService = nil
+  }
+  
+  public func loadCachedData() async {
+    await withTaskGroup(of: Void.self) { group in
+      group.addTask {
+        await self.loadCachedFilterLists()
+      }
+      group.addTask {
+        await self.loadCachedDefaultFilterList()
+      }
+    }
+  }
+  
+  private func loadCachedFilterLists() async {
+    let settingsInfo: [(index: Int, uuid: String, folderURL: URL?, resources: [ResourceDownloader.Resource])] = await MainActor.run {
```

Comment:
In the future lets use structs instead of tuples larger than 3 things, it gets pretty messy